keps/0009-operator-toolkit.md

```diff
@@ -0,0 +1,394 @@
+
+## Goals
+
+The toolkit must provide the ability for:
+
+- conditional inclusions
+- defining default values
+- replacement of values
+- removal of keys and values
+
+## Proposal
+
+A KUDO framework is split into several base components, ordered from the outside-in:
+
+- [Plans](#plans)
+- [Steps](#steps)
+- [Tasks](#tasks)
+- [Templates](#templates)
+- [Parameters](#parameters)
+- [Extensions and Bases](#extensions-and-bases)
+- [Execution State](#execution-state)
+
+These combine to form an framework responsible for the deployment, upgrading, and day 2 operations of software deployed on Kubernetes. While KUDO ships with a templating system, it is not intended to advance the state of deploying software on Kubernetes, and provides facilities for integrating with other templating and deployment systems. KUDO's focus is instead on the sequencing and day 2 operations of this software, and is able to take advantage of existing templating to fulfill this need.
+
+### Definitions
+
+### Framework Organization
+
+An framework bundle is a folder that contains all of the manifests needed to create or extend a KUDO framework. In the most basic form, an framework bundle is structured in the following format:
+
+```shell
+.
+├── framework.yaml
+├── params.yaml
+└── common
+    └── common.yaml
+└── templates
+    ├── backup.yaml
+    ├── clear-data.yaml
+    ├── deployment.yaml
+    ├── init.yaml
+    ├── load-data.yaml
+    ├── pvc.yaml
+    ├── restore.yaml
+    └── service.yaml
+```
+
+#### framework.yaml
+
+`framework.yaml` is the base definition of an framework. It follows the following format, extracted from the MySQL example:
+
+```yaml
+name: "my-framework"
+description: A test example framework.
+version: "5.7"
+kudoVersion: ">= 0.2.0"
+kubeVersion: ">= 1.14"
+maintainers:
+  - Bob <bob@example.com>
+  - Alice <alice@example.com>
+url: https://github.com/myframework/myframework
+tasks:
+  deploy:
+    resources:
+      - pvc.yaml
+      - deployment.yaml
+      - service.yaml
+      - deployment2.yaml
+      - job.yaml
+    patches:
+      - deploy-patch.yaml
+    patchesStrategicMerge:
+      - super-weird-deploy-patch.yaml
+  init:
+    resources:
+      - init.yaml
+  pv:
+    resources:
+      - pvc.yaml
+  backup:
+    resources:
+      - init.yaml
+    patches:
+      - backup.yaml
+  restore:
+    resources:
+      - init.yaml
+    patches:
+      - restore.yaml
+  load-data:
+    resources:
+      - init.yaml
+    patches:
+      - load-data.yaml
+  clear-data:
+    resources:
+      - init.yaml
+    patches:
+      - clear-data.yaml
+  query:
+    resources:
+      - job.yaml
+plans:
+  deploy:
+    steps:
+      - name: deploy
+        tasks:
+          - deploy
+      - name: init
+        tasks:
+          - init
+      - name: cleanup
+        tasks:
+          - init
+        delete: true
+  backup:
```

Comment:
What is the difference if any between backup (the plan) and backup (the task) ?

Reply:
They're both using the same task definition (set of kustomize resources + patches), but have a different operation (DELETE request instead of patch/server side apply)